feat(#23): Add metrics

randreev1321 opened the pull request:
Changes:
- Add collect and export backups metrics to tmp file for using in server mode.
- Add api-server module to to export metrics for Prometheus.
- Add new command `server` to start nxs-backup in server mode.
- Add new config options to set server parameters.

Refs: #23

randreev1321 committed:
feat(#23): Add metrics

Changes:
- Add collect and export backups metrics to tmp file for using in server mode.
- Add api-server module to to export metrics for Prometheus.
- Add new command `server` to start nxs-backup in server mode.
- Add new config options to set server parameters.

Refs: #23

borisershov commented on ctx/jobs_init.go:
```diff
@@ -76,10 +79,6 @@ func jobsInit(conf ConfOpts, storages map[string]interfaces.Storage) ([]interfac
-		//}
-
 		switch j.JobType {
 		case misc.AllowedJobTypes[0]:
```
May be better to use constants with appropriate type?

randreev1321 replied:
Agreed. Corrected.

borisershov commented on ctx/jobs_init.go:
```diff
@@ -22,12 +22,15 @@ import (
-func jobsInit(conf ConfOpts, storages map[string]interfaces.Storage) ([]interfaces.Job, error) {
-	var errs *multierror.Error
-	var jobs []interfaces.Job
+func jobsInit(conf ConfOpts, storages map[string]interfaces.Storage, metricsData *metrics.Data) ([]interfaces.Job, error) {
```
Would be better to use struct with settings to transfer args

randreev1321 replied:
Agreed. Corrected.

borisershov commented on interfaces/job.go:
```diff
@@ -5,6 +5,8 @@ import (
 )
 
 type Job interface {
+	FillMetrics(ofs string, metrics map[string]float64)
```
Set or Add?

randreev1321 replied:
Agreed. Corrected.

borisershov commented on modules/metrics/metrics.go:
```diff
@@ -0,0 +1,75 @@
+		err error
+	)
+
+	retry := 3
```
Use const for this cases

randreev1321 replied:
Agreed. Corrected.

randreev1321 committed:
feat(#23): Review fixes and improvements

Changes:
- Rename functions.
- Use constants for allowed backup types.
- Options structs for init functions.
- Change metrics store logic. Reuse previous metrics if job not run.

Refs: #23

borisershov commented on interfaces/storage.go:
```diff
@@ -77,7 +87,7 @@ func (s Storages) CleanupTmpData(job Job) error {
 
 		tmpBakFile := dumpObj.TmpFile
-		if job.GetType() == misc.IncBackupType {
+		if job.GetType() == string(misc.IncFiles) {
```
`job.GetType()` must return type `misc.BackupType`

borisershov commented on misc/generals.go:
```diff
@@ -17,29 +17,43 @@ import (
-	"external",
-}
+const (
+	DescFiles            BackupType = "desc_files"
```
This  consts need to have a prefix (e.g. `BackupType`) to be unique within the `misc`.

borisershov commented on modules/backup/desc_files/desc_files.go:
```diff
@@ -184,15 +227,26 @@ func (j *job) DoBackup(logCh chan logger.LogRecord, tmpDir string) error {
 		}
+		fileInfo, _ := os.Stat(tmpBackupFile)
+		j.SetOfsMetrics(ofsPart, map[string]float64{
+			"backup_ok":   float64(1),
```
This must be a defined constants to prevent syntax errors

borisershov commented on modules/backup/external/external.go:
```diff
@@ -31,11 +35,13 @@ type JobParams struct {
 	SafetyBackup     bool
 	SkipBackupRotate bool
 	Storages         interfaces.Storages
+	Metrics          *metrics.Data
```
What is the difference between the `Metrics` and `OldMetrics`?

randreev1321 committed:
feat(#23): Review fixes and improvements

randreev1321 committed:
feat(#23): Docs update

randreev1321 committed:
feat(#23): Update metrics names
